README.md
# Node2Neo

A Node.js transaction REST API for neo4j.
Node2neo uses the transaction endpoints feature of the Neo4j REST API. This reduces the traffic between the server and neo4j and allows for multiple http requests within the same transaction.

This is intended to be a building block for future modules to add advanced functionality.

NOTE: Neo4j 2.0 is required.

## Installation

    npm install node2neo

## Usage

#### Connecting to the db

```js
var db = require('node2neo').db('http://localhost:7474');
```

There are only five methods offered by node2neo:
1. beginTransaction
2. executeStatement
3. commitTransaction
4. removeTransaction
5. getTransactionId (helper)

#### Begin Transaction
To create a transaction use the db.beginTransaction function. The statements provided to the beginTransaction function must match the required neo4j format.

````js
var transStatement = {
  statements: [{
    statement: 'CREATE n:User RETURN n'
  }, {
    statement: 'CREATE o:Person RETURN o'
  }]
}

db.beginTransaction(transStatement, function(err, results){

})
```

You can immediately commit a transaction by passing an options variable with a commmit property

```js
db.beginTransaction(transStatement, { commit: true }, function(err, results){

})
```

The response format for an open transaction is as follows (closed transaction have a different response)

```js
response: {
  results: [{
    commit: '...commit string...',
    transaction: {expires: '...transaction timeout period...'},
    errors: []
    results: [{
      columns: ['n', 'o'],
      data: [{
        row: [{ name: 'MyNode'}, {name: 'YourNode'}]
      }]
    }]

  }]
}
```

#### Execute Statement
To perform another action in the same transaction use the execute Statement function

In order to execute a statement you will need the transaction Id or commit string.
You can use the db.getTransactionId(commitString) helper to get the transaction id from the commit string or just pass in the commit string

```js
db.executeStatement(commit, statements, function(err, results){

})
```

Transaction will expire after  aperiod of time. You can prevent a transaction from expiring by submitting an empty statements array

```js
var statemetns = {
  statements: []
}
db.executeStatement(commit, statements, function(err, results){

})
```

#### Commit a transaction
All of the statements in a transaction are not commited until the whole transaction is committed. You can optionally include some further statements at this point before committing the transaction. The result from a commit is different:


```js
db.commitTransaction(commit, [statements], function(err, results){

})

// if no statement is provided you receive tw empy arrays.
// If a statement was provided the results array would be poopulated as above
{
  "results" : [ ],
  "errors" : [ ]
}
```

#### Remove a transaction
There are two ways to undo a transaction that is in mid progress: let it timeout or tell Neo4j to remove it.

```js
db.removeTransaction(commit, function(err){

})
```

##Licence
MIT

CHANGELOG.md
##0.0.8 (2014-04-14)
 - Added console log on database connect

## 0.0.7 (2014-04-13)
 - Ability to create a blank transaction

## 0.0.6 (2014-02-02)
 - Fixed dependencies

 ## 0.0.5 (2014-02-01)
 - Added streams

## 0.0.4 (2013-11-12)
  - Added error handling for Connection Refused

## 0.0.3 (2013-10-11)

Features:

  - Removed transactions. Moved to node2neo-transactions module
  - Reverted the require syntax to require('node2neo')(url)

Bugfixes:

## 0.0.2 (2013-10-11)

Features:

  - Added transactions. require('node2neo').transaction
  - Changed the the require syntax to require('node2neo').db(url)

Bugfixes:


## 0.0.1 (2013-10-09)

  - Initial commit

test/node2neo.js
'use strict';

// var testDatabase = require('./util/database');
// var db = require('../')(testDatabase.url);
var db = require('../')('http://localhost:7475');
// var nock = require('nock')('http://localhost:' + testDatabase.port);

var should = require('chai').should();

describe('node2neo', function () {

  // before(testDatabase.refreshDb);
  // after(testDatabase.stopDb);

  var commit, transaction, transId, ids = [];

  // it('should create a transaction: pipe', function (done) {
  //   var statements = {
  //     statements: [{
  //       statement: 'CREATE (n:User {name: \'Rory\' }) RETURN id(n)'
  //     }]
  //   };

  //   //prepare for input
  //   // var ws = Writable({ objectMode: true });
  //   // ws._write = function (chuck, enc, next) {

  //   // }

  //   var results = db.beginTransactionStream(statements);
  //   results.on('data', function (results) {
  //     console.log(results.toString());
  //     transId = db.getTransactionId(results.commit);
  //     commit = results.commit;
  //     transaction = results.transaction;
  //     should.exist(results.results);
  //     should.exist(results.transaction);
  //     results.errors.should.be.an('array');
  //     results.errors.length.should.equal(0);

  //     results.results.length.should.equal(1);
  //     results.results[0].columns.length.should.equal(1);
  //     results.results[0].columns[0].should.equal('id(n)');
  //     results.results[0].data.length.should.equal(1);
  //     var id = results.results[0].data[0].row[0];
  //     ids.push(id);

  //     // validate that the record hasn't been created in the database
  //     db.beginTransaction({statements: [{statement: 'START n = node(' + id + ') RETURN n'}]}, {commit: true},
  //       function (err, results) {
  //       should.not.exist(err);
  //       should.exist(results.results);
  //       Object.keys(results.results[0].data[0].row[0]).length.should.equal(0);
  //     });
  //   });
  //   results.on('end', done);
  // });

  // it('should error on invalid statement: pipe', function (done) {
  //   var results = db.beginTransactionStream('blue');
  //     console.log(results);
  //   results.on('error', function (error) {
  //     error.code.should.equal('Neo.ClientError.Request.InvalidFormat');
  //     done();
  //   });
  // });

  // it('should error on database error', function (done) {

  //   nock
  //     .filteringRequestBody(function () {
  //       return '*';
  //     })
  //     .post('/db/data/transaction', '*')
  //     .reply(500, 'Oops');

  //   db.beginTransactionStream('blue')
  //     .on('error', function (error) {
  //       error.code.should.equal('Neo.ClientError.Request.InvalidFormat');
  //       done();
  //     })
  //     .on('data', function (data) {
  //       console.log(data);
  //     });
  // });

  // it('should create a stream query', function (done) {
  //   var statements = {
  //     statements: [{
  //       statement: 'CREATE (n:User {name: \'Rory\' }) RETURN id(n)'
  //     }]
  //   };

  //   var results = db.query(statements);
  //   results.on('data', function (results) {
  //     results.should.be.an.instanceof(Array);
  //     var id = results[0].row[0];

  //     // validate that the record has been created in the database
  //     db.query({statements: [{statement: 'START n = node(' + id + ') RETURN n'}]})
  //       .on('data', function (data) {
  //         var node = data[0].row[0];

  //         node.name.should.equal('Rory');
  //       })
  //       .on('end', done);
  //   });
  // });

  it('should create a transaction', function (done) {
    var statements = {
      statements: [{
        statement: 'CREATE (n:User {name: \'Rory\' }) RETURN id(n)'
      }]
    };
    db.beginTransaction(statements, function (err, results) {
      should.not.exist(err);
      transId = db.getTransactionId(results.commit);
      commit = results.commit;
      transaction = results.transaction;
      should.exist(results.results);
      should.exist(results.transaction);
      results.errors.should.be.an('array');
      results.errors.length.should.equal(0);

      results.results.length.should.equal(1);
      results.results[0].columns.length.should.equal(1);
      results.results[0].columns[0].should.equal('id(n)');
      results.results[0].data.length.should.equal(1);
      var id = results.results[0].data[0].row[0];
      ids.push(id);

      // validate that the record hasn't been created in the database
      db.beginTransaction({statements: [{statement: 'START n = node(' + id + ') RETURN n'}]}, {commit: true},
        function (err, results) {
        should.not.exist(err);
        should.exist(results.results);
        Object.keys(results.results[0].data[0].row[0]).length.should.equal(0);
        done();
      });
    });
  });

  it('should create a blank transaction', function (done) {
    db.beginTransaction(function (err, results) {
      should.not.exist(err);
      should.exist(results.results);
      should.exist(results.transaction);
      results.errors.should.be.an('array');
      results.errors.length.should.equal(0);
      results.results.length.should.equal(0);

      // validate that the record hasn't been created in the database
      db.removeTransaction(db.getTransactionId(results.commit),
        function (err) {
        should.not.exist(err);
        done();
      });
    });
  });

  it('should add another statement to a transaction', function (done) {
    var statements = {
      statements: [{
        statement: 'CREATE (o:Person {name:\'Cath\'}) RETURN id(o)'
      }, {
        statement: 'CREATE (p:Person { name: \'Philip\' }) RETURN p.name, id(p)'
      }]
    };

    db.executeStatement(transId, statements, function (err, results) {
      should.not.exist(err);
      should.exist(results.results);
      should.exist(results.transaction);
      db.getTransactionId(results.commit).should.equal(transId);
      results.errors.should.be.an('array');
      results.errors.length.should.equal(0);

      results.results.length.should.equal(2);
      results.results[0].columns.length.should.equal(1);
      results.results[0].columns[0].should.equal('id(o)');
      results.results[0].data.length.should.equal(1);
      var id1 = results.results[0].data[0].row[0];
      ids.push(id1);
      results.results[1].columns.length.should.equal(2);
      results.results[1].columns[0].should.equal('p.name');
      results.results[1].columns[1].should.equal('id(p)');
      results.results[1].data.length.should.equal(1);
      results.results[1].data[0].row[0].should.equal('Philip');
      var id2 = results.results[1].data[0].row[1];
      ids.push(id2);

      // validate that the record hasn't been created in the database
      db.beginTransaction({statements: [{statement: 'START n = node(' + id1 + '), o = node(' + id2 + ') RETURN n, o'}]},
        { commit: true }, function (err, results) {
        should.not.exist(err);
        should.exist(results.results);
        Object.keys(results.results[0].data[0].row[0]).length.should.equal(0);
        done();
      });
    });
  });

  it('should add another statement to a transaction: commit string', function (done) {
    var statements = {
      statements: [{
        statement: 'CREATE (z:Person {name:\'Zorg\'}) RETURN id(z)'
      }]
    };
    db.executeStatement(commit, statements, function (err, results) {
      should.not.exist(err);
      should.exist(results.results);
      done();
    });
  });

  it('should extend the timelimit of a transaction', function (done) {
    db.executeStatement(transId, function (err, results) {
      should.not.exist(err);
      transaction.expires.should.be.below(results.transaction.expires);
      done();
    });
  });

  it('should commit a transaction - with no additional statement', function (done) {
    db.commitTransaction(transId, function (err, results) {
      should.not.exist(err);
      should.exist(results.results);
      should.not.exist(results.transaction);
      results.errors.should.be.an('array');
      results.errors.length.should.equal(0);
      results.results.length.should.equal(0);

      // validate that the records have been created in the database
      var statement = [];
      for (var i = 0, len = ids.length; i < len; i++) {
        statement.push({statement: 'START n' + i + '=node(' + ids[i] + ') RETURN n' + i });
      }
      var statements = {
        statements: statement
      };
      db.beginTransaction(statements, {commit: true}, function (err, results) {
        should.not.exist(err);
        should.exist(results.results);
        done();
      });
    });
  });

  it('should create a transaction, and immediately commit', function (done) {
    var statements = {
      statements: [{
        statement: 'CREATE (n:User {name: \'Rory-again\' }) RETURN id(n)'
      }]
    };
    db.beginTransaction(statements, {commit: true}, function (err, results) {
      should.not.exist(err);
      should.exist(results.results);
      results.errors.should.be.an('array');
      results.errors.length.should.equal(0);

      results.results.length.should.equal(1);
      results.results[0].columns.length.should.equal(1);
      results.results[0].columns[0].should.equal('id(n)');
      results.results[0].data.length.should.equal(1);

      var id = results.results[0].data[0].row[0];

      // validate that the record hasn't been created in the database
      db.beginTransaction({statements: [{statement: 'START n = node(' + id + ') RETURN n'}]}, {commit: true},
        function (err, results) {
        should.not.exist(err);
        should.exist(results.results);
        results.results[0].data[0].row[0].name.should.equal('Rory-again');
        done();
      });
    });
  });

  it('should commit a transaction with an additional statement', function (done) {
    var statements1 = {
      statements: [{
        statement: 'CREATE (n:User {name: \'Rory-third\' }) RETURN id(n)'
      }]
    };
    var statements2 = {
      statements: [{
        statement: 'CREATE (n:User {name: \'Cath-again\' }) RETURN id(n)'
      }]
    };
    db.beginTransaction(statements1, function (err, results) {
      should.not.exist(err);
      transId = db.getTransactionId(results.commit);
      db.commitTransaction(transId, statements2, function (err, results) {
        should.not.exist(err);
        should.exist(results.results);
        should.not.exist(results.transaction);
        results.errors.should.be.an('array');
        results.errors.length.should.equal(0);

        results.results.length.should.equal(1);
        results.results[0].columns.length.should.equal(1);
        results.results[0].columns[0].should.equal('id(n)');
        results.results[0].data.length.should.equal(1);
        var id = results.results[0].data[0].row[0];

        // validate that the record hasn't been created in the database
        db.beginTransaction({statements: [{statement: 'START n = node(' + id + ') RETURN n'}]}, {commit: true},
          function (err, results) {
          should.not.exist(err);
          should.exist(results.results);
          results.results[0].data[0].row[0].name.should.equal('Cath-again');
          done();
        });
      });
    });
  });


  it('should rollback a transaction', function (done) {
    var statements1 = {
      statements: [{
        statement: 'CREATE (n:User {name: \'Cath-third\' }) RETURN id(n)'
      }]
    };
    db.beginTransaction(statements1, function (err, results) {
      should.not.exist(err);
      transId = db.getTransactionId(results.commit);
      var id = results.results[0].data[0].row[0];
      db.removeTransaction(transId, function (err) {
        should.not.exist(err);

        // validate that the record hasn't been created in the database
        db.beginTransaction({statements: [{statement: 'START n = node(' + id + ') RETURN n'}]}, {commit: true},
          function (err, results) {
          err[0].code.should.equal('Neo.ClientError.Statement.EntityNotFound');
          err[0].message.should.contain('Node with id ');
          should.not.exist(results);
          done();
        });
      });
    });
  });

  it('should error on bad input:beginTransaction', function (done) {
    db.beginTransaction({statements: [{statement: 'bad cypher syntax'}]}, {commit: true},
      function (err, results) {
      err[0].code.should.equal('Neo.ClientError.Statement.InvalidSyntax');
      // err[0].status.should.equal('STATEMENT_SYNTAX_ERROR');
      should.not.exist(results);
      // statements is not an array
      db.beginTransaction({statements: {statement: 'CREATE n RETURN n'}}, {commit: true},
        function (err, results) {
        err[0].code.should.equal('Neo.ClientError.Request.InvalidFormat');
        // err[0].status.should.contain('INVALID_REQUEST_FORMAT');
        should.not.exist(results);
        done();
      });
    });
  });

  it('should error on bad input:executeStatement', function (done) {
    db.beginTransaction({statements: [{statement: 'CREATE n RETURN n'}]}, function (err, results) {
      should.not.exist(err);
      transId = db.getTransactionId(results.commit);
      db.executeStatement(transId, {statements: [{statement: 'bad syntax'}]}, function (err, results) {
        err[0].code.should.equal('Neo.ClientError.Statement.InvalidSyntax');
        // err[0].status.should.equal('STATEMENT_SYNTAX_ERROR');
        should.not.exist(results);
        // statements is not an array
        db.beginTransaction({statements: {statement: 'CREATE n RETURN n'}}, {commit: true},
          function (err, results) {
          err[0].code.should.equal('Neo.ClientError.Request.InvalidFormat');
          // err[0].status.should.contain('INVALID_REQUEST_FORMAT');
          should.not.exist(results);
          done();
        });
      });
    });
  });

  it('should error on bad input:removeTransaction', function (done) {
    db.removeTransaction(28, function (err) {
      err[0].code.should.equal('Neo.ClientError.Transaction.UnknownId');
      // err[0].status.should.equal('INVALID_TRANSACTION_ID');
      done();
    });
  });

  // it('should return results in a graph format', function (done) {
  //   db.beginTransaction({statements: [{statement: 'MATCH n RETURN n'}], resultDataContents: [ 'graph' ]},
  //     {commit: true}, function (err, results) {
  //     should.not.exist(err);
  //     should.exist(results.results[0].data[0].graph);
  //     done();
  //   });
  // });
});

test/transaction.js
var testDatabase = require('./util/database');
var db = require('../').db(testDatabase.url);
var Transaction = require('../').transaction(db);
var EventEmitter = require('events').EventEmitter;

var should = require('chai').should();



var transaction, User, rootUserId, userData;


describe("transaction", function(){
  // before(testDatabase.refreshDb);
  // after(testDatabase.stopDb);
  it("should begin a transaction", function(done){
    Transaction.begin(function(err, response){
      should.exist(Transaction._commit);
      Transaction._statements.should.be.an('array').with.length(0);
      Transaction.commit(function(err2, response2){
        //test the begin
        should.not.exist(err);
        should.exist(response);
        response.length.should.be.equal(0);

        //test the commit
        should.not.exist(err2);
        should.not.exist(Transaction._commit);
        Transaction._statements.should.be.an('array').with.length(0);
        should.exist(response2);
        response2.length.should.be.equal(0);
        done();
      });
    });
  });
  it("should begin a Transaction: with string statement", function(done){
    var statement = 'CREATE n:User RETURN id(n)';
    Transaction.begin(statement, function(err, response){
      should.exist(Transaction._commit);
      Transaction._appliedStatements.should.be.an('array').with.length(1);
      Transaction.commit(function(err2, response2){
        //test teh begin
        should.not.exist(err);
        should.exist(response);
        response.length.should.be.equal(1);
        response[0].columns[0].should.be.equal('id(n)');
        response[0].data[0].row[0].should.be.a('number');

        //test the commit
        should.not.exist(err2);
        should.not.exist(Transaction._commit);
        Transaction._statements.should.be.an('array').with.length(0);
        should.exist(response2);
        response2.length.should.be.equal(0);
        done();
      });
    });
  });
  it("should begin a Transaction: with object statement", function(done){
    var statement = {
      statement: 'CREATE n:User RETURN id(n)'
    };
    Transaction.begin(statement, function(err, response){
      should.exist(Transaction._commit);
      Transaction._appliedStatements.should.be.an('array').with.length(1);
      Transaction.commit(function(err2, response2){
        // test the begin
        should.not.exist(err);
        should.exist(response);
        response.length.should.be.equal(1);
        response[0].columns[0].should.be.equal('id(n)');
        response[0].data[0].row[0].should.be.a('number');

        //test the commit
        should.not.exist(err2);
        should.not.exist(Transaction._commit);
        Transaction._statements.should.be.an('array').with.length(0);
        should.exist(response2);
        response2.length.should.be.equal(0);
        done();
      });
    });
  });
  it("should error if trying to open a second transaction", function(){
    // Add your behavior testing code here
  });
});

test/util/database.js
var TEST_INSTANCE_PORT = parseInt(process.env.TEST_INSTANCE_PORT || '10506', 10);
var disposableSeraph = require('disposable-seraph');

var _nsv;

var refreshDb = function(done) {
  disposableSeraph({
      version: '2.0.0',
      edition: 'community',
      port: TEST_INSTANCE_PORT
    },
    function(err, _, nsv) {
      _nsv = nsv;
      done(err);
    });
};

var stopDb = function(done) {
  _nsv.stop(done);
};

module.exports = {
  port: TEST_INSTANCE_PORT,
  url: 'http://localhost:' + TEST_INSTANCE_PORT,
  refreshDb: refreshDb,
  stopDb: stopDb
};

lib/errors.js
'use strict';

var util = require('util');

/**
 * Application wide error codes.
 * @type {ErrorCodes}
 */

// From http://dustinsenos.com/articles/customErrorsInNode
// Create a new Abstract Error constructor
var AbstractError = function (code, msg, constr) {
  // If defined, pass the constr property to V8's
  // captureStackTrace to clean up the output
  Error.call(this);
  Error.captureStackTrace(this, constr || this);

  // If defined, store a custom error message
  this.message = msg || 'Error';

  // if defined store the code
  this.code = code;
};
util.inherits(AbstractError, Error);
AbstractError.prototype.name = 'Abstract Error';



// Create the ApiError
var CypherError = function (msg) {
  var code = msg.code;
  msg = msg.msg;
  CypherError.super_.call(this, code, msg, this.constructor);
};
util.inherits(CypherError, AbstractError);
CypherError.prototype.name = 'Cypher Error';



// Create the DatabaseError
var DatabaseError = function (msg) {
  var code = null;
  DatabaseError.super_.call(this, code, msg, this.constructor);
};
util.inherits(DatabaseError, AbstractError);
DatabaseError.prototype.name = 'Database Error';

// Create the JSONError
var JSONError = function (msg) {
  var code = null;
  JSONError.super_.call(this, code, msg, this.constructor);
};
util.inherits(JSONError, AbstractError);
JSONError.prototype.name = 'JSON Error';

module.exports = {
  CypherError: CypherError,
  DatabaseError: DatabaseError,
  JSONError: JSONError
};

lib/index.js
'use strict';

var request = require('request');
var through = require('through');
var JSONStream = require('JSONStream');
var Errors = require('./errors');


// helper function to check if a value is a number
function isNumber(o) {
  return ! isNaN(o - 0) && o !== null && o !== '' && o !== false;
}

/**
 * Instantiate the Node2Neo object
 * @param {string} server The url fo the server e.g. http://localhost:7474
 */
function Node2Neo(server) {
  if (!server) {
    server = 'http://localhost:7474';
  }
  this.server = server.replace(/\/$/, ''); // remove trailing /

  // just run this async -  no need to block the program
  var req = {
    url: this.server + '/db/data',
    headers: {
      'Accept': 'application/json',
      'X-Stream': true
    }
  };
  var self = this;
  request.get(req, function (err, res, body) {
    if (err) {
      console.log('Error connecting to the Neo4j database at ' + self.server);
    }
    else {
      try {
        var graph = JSON.parse(body);
        console.log('Successfully connected to the Neo4j v.' + graph.neo4j_version + ' Instance at ' + self.server);
      }
      catch (e) {
        console.log('Error parsing response from Neo4j: ' + body);
      }

    }
  });
}

/**
 * Begin a http transaction to Neo4j
 * var db = require('node2neo')('http://localhost:7474')
 * var statements = {
 *   "statements" : [ {
 *     "statement" : "CREATE n RETURN id(n)"
 *   } ]
 * }
 * db.beginTransaction(statements, cb);
 *
 * Execute the transaction immediately
 * var options = {
 *  commit: true
 * }
 * db.beginTransaction(statements, options, cb);
 *
 *
 * @param  {object}   statements The statements object as defined in http://docs.neo4j.org/chunked/milestone/rest-api-transactional.html.
 *                               Contains an array of objects, each containing a statement variable with a cypher statement.
 *                               Optionally may contain a resultDataContents array. Options include row and graph.
 * @param  {object}   options    Options object. Set Options.commit to true to commit the transaction immediately.
 * @param  {Function} callback
 * @return {Object}              Contains a results array and an errors array.
 */
Node2Neo.prototype.beginTransaction = function (statements, options, callback) {

  if (typeof options === 'function') {
    callback = options;
  }
  if (typeof statements === 'function') {
    callback = statements;
  }

  var commit = '';
  if (options && options.commit) commit = '/commit';

  var req = {
    url: this.server + '/db/data/transaction' + commit,
    headers: {
      'Accept': 'application/json',
      'X-Stream': true
    }
  };

  if (typeof statements !== 'function') {
    req.json = statements;
  }

  request.post(req,
  function (err, res, body) {
    if (err) {
      return callback(err);
    }
    if (typeof body === 'string') {
      body = JSON.parse(body);
    }
    if (body.errors && body.errors.length > 0) {
      return callback(body.errors);
    }

    else return callback(null, body);
  });
};



Node2Neo.prototype.beginTransactionStream = function (statements, options) {

  if (typeof statements === 'undefined') {
    return new Error('You need to provide at least one cypher statement');
  }

  var commit = '';
  if (options && options.commit) commit = '/commit';

  return request
    .post({
      url: this.server + '/db/data/transaction' + commit,
      json: statements,
      headers: {
        'Accept': 'application/json',
        'X-Stream': true
      }
    })
    .on('response', function (res) {
      var self = this;
      if (!(res.statusCode >= 200 && res.statusCode < 300)) {
        self.emit('error', new Errors.DatabaseError('Got response with HTTP code ' + res.statusCode));
      }
      else {
        res.pipe(JSONStream.parse())
          .on('error', function (err) {
            self.emit('error', new Errors.JSONError(err));
          })
          .pipe(through(function (data) {
            if (data.errors.length > 0) {
              self.emit('error', new Errors.CypherError(data.errors[0]));
            }
            else {
              self.emit('data', data);
            }
          }));
      }
    });
};


Node2Neo.prototype.query = function (statements) {

  if (typeof statements === 'undefined') {
    return new Error('You need to provide at least one cypher statement');
  }

  return request
    .post({
      url: this.server + '/db/data/transaction/commit',
      json: statements,
      headers: {
        'Accept': 'application/json',
        'X-Stream': true
      }
    })
    .on('error', function (err) {
      return new Errors.DatabaseError(err);
    })
    .pipe(JSONStream.parse())
    .on('error', function (err) {
      return new Errors.JSONError(err);
    })
    .pipe(through(function (data) {
      if (data.errors.length > 0) {
        this.emit('error', data.errors[0]);
      }
      else {
        this.emit('data', data.results[0].data);
      }
    }))
    .on('error', function (err) {
      return new Errors.CypherError(err);
    });
};


/**
 * Execute an additional statement on an existing transaction.
 *
 * Statements have the same syntax as creating a new transaction.
 *
 * @param  {Number/String}   transId    The id / commit string of the transaction.
 * @param  {Object}   statements The statement to be executed.
 * @param  {Function} callback
 * @return {Object}              Contains a results array and an errors array.
 */
Node2Neo.prototype.executeStatement = function (transId, statements, callback) {
  // just resetting the timeout on the transaction
  if (typeof statements === 'function') {
    callback = statements;
    // an empty statements block is used to extend the transaction time
    statements = {
      statements: []
    };
  }

  transId = this.getTransactionId(transId);

  if (!isNumber(transId)) {
    return callback(new Error('You must provide a numeric transaction id.'));
  }

  // request
  //   .post(this.server + '/db/data/transaction/' + transId)
  //   .send(statements)
  //   .set('Accept', 'application/json')
  //   .set('X-Stream', true)
  //   .end(function (err, res) {
  //     if (err) {
  //       return callback(err);
  //     }
  //     if (res.body.errors.length > 0) {
  //       return callback(res.body.errors);
  //     }

  //     else return callback(null, res.body);
  //   });

  request.post({
    url: this.server + '/db/data/transaction/' + transId,
    json: statements,
    headers: {
      'Accept': 'application/json',
      'X-Stream': true
    }
  },
  function (err, res, body) {
    if (err) {
      return callback(err);
    }
    if (body.errors && body.errors.length > 0) {
      return callback(body.errors);
    }

    else return callback(null, body);
  });
};


/**
 * Commit an open transaction, performing all of the statements.
 * You can optionally add additional statemetns at this point as well.
 *
 * @param  {Number/String}   transId    The id / commit string of the transaction.
 * @param  {Object}   statements The statement to be executed.
 * @param  {Function} callback   [description]
 * @return {Object}              Contains a results array and an errors array.
 */
Node2Neo.prototype.commitTransaction = function (transId, statements, callback) {
  if (typeof statements === 'function') {
    callback = statements;
    statements = undefined;
    // return callback(new Error('You need to provide at least one cypher statement'));
  }
  transId = this.getTransactionId(transId);

  if (!isNumber(transId)) {
    return callback(new Error('You must provide a numeric transaction id.'));
  }

  // request
  //   .post(this.server + '/db/data/transaction/' + transId + '/commit')
  //   .send(statements)
  //   .set('Accept', 'application/json')
  //   .set('X-Stream', true)
  //   .end(function (err, res) {
  //     if (err) {
  //       return callback(err);
  //     }
  //     if (res.body.errors.length > 0) {
  //       return callback(res.body.errors);
  //     }

  //     else return callback(null, res.body);
  //   });
  request.post({
    url: this.server + '/db/data/transaction/' + transId + '/commit',
    json: statements || true,
    headers: {
      'Accept': 'application/json',
      'X-Stream': true
    }
  },
  function (err, res, body) {
    if (err) {
      return callback(err);
    }
    if (body.errors && body.errors.length > 0) {
      return callback(body.errors);
    }

    else return callback(null, body);
  });
};


/**
 * Remove an open transaction
 * @param  {Number/String}   transId    The id / commit string of the transaction.
 * @param  {Function} callback
 * @return {Null}
 */
Node2Neo.prototype.removeTransaction = function (transId, callback) {
  transId = this.getTransactionId(transId);
  if (!isNumber(transId)) {
    return callback(new Error('You must provide a numeric transaction id.'));
  }
  // request
  //   .del(this.server + '/db/data/transaction/' + transId)
  //   .set('Accept', 'application/json')
  //   .set('X-Stream', true)
  //   .end(function (err, res) {
  //     if (err) {
  //       return callback(err);
  //     }
  //     if (res.body.errors.length > 0) {
  //       return callback(res.body.errors);
  //     }

  //     else return callback(null, res.status);
  //   });
  request.del({
    url: this.server + '/db/data/transaction/' + transId,
    headers: {
      'Accept': 'application/json',
      'X-Stream': true
    },
    json: true
  },
  function (err, res, body) {
    if (err) {
      return callback(err);
    }
    if (body.errors && body.errors.length > 0) {
      return callback(body.errors);
    }

    else return callback(null, body);
  });
};

Node2Neo.prototype.getTransactionId = function (commit) {
  if (isNumber(commit)) return commit;
  return commit.slice(commit.indexOf('/transaction') + 13, -7);
};

// export variables

module.exports = function (db) {
  return new Node2Neo(db);
};
